x-pack/plugins/security_solution/server/endpoint/services/agent/clients/crowdstrike/crowdstrike_agent_status_client.ts

```diff
@@ -0,0 +1,171 @@
+  UNKNOWN = 'unknown',
+}
+
+export class CrowdstrikeAgentStatusClient extends AgentStatusClient {
+  protected readonly agentType: ResponseActionAgentType = 'crowdstrike';
+  private async getAgentStatusFromConnectorAction(agentIds: string[]) {
+    const connectorActions = new NormalizedExternalConnectorClient(
+      this.options.connectorActionsClient,
+      this.log
+    );
+    connectorActions.setup(CROWDSTRIKE_CONNECTOR_ID);
+
+    const agentStatusResponse = (await connectorActions.execute({
+      params: {
+        subAction: SUB_ACTION.GET_AGENT_ONLINE_STATUS,
+        subActionParams: {
+          ids: agentIds,
+        },
+      },
+    })) as ActionTypeExecutorResult<CrowdstrikeGetAgentOnlineStatusResponse>;
+
+    return keyBy(agentStatusResponse.data?.resources, 'id');
+  }
+  async getAgentStatuses(agentIds: string[]): Promise<AgentStatusRecords> {
+    const esClient = this.options.esClient;
+    const metadataService = this.options.endpointService.getEndpointMetadataService();
+    const sortField = 'crowdstrike.host.last_seen';
+
+    const query = {
+      bool: {
+        must: [
+          {
+            bool: {
+              filter: [
+                {
+                  terms: {
+                    'crowdstrike.host.id': agentIds,
+                  },
+                },
+              ],
+            },
+          },
+        ],
+      },
+    };
+
+    try {
+      const [searchResponse, allPendingActions] = await Promise.all([
+        esClient.search(
+          {
+            index: CROWDSTRIKE_AGENT_INDEX_PATTERN,
+            from: 0,
+            size: 10000,
+            query,
+            collapse: {
+              // TODO: check if we should use crowdstrike.cid instead
+              field: 'crowdstrike.host.id',
+              inner_hits: {
+                name: 'most_recent',
+                size: 1,
+                sort: [
+                  {
+                    [sortField]: {
+                      order: 'desc',
+                    },
+                  },
+                ],
+              },
+            },
+            sort: [
+              {
+                [sortField]: {
+                  order: 'desc',
+                },
+              },
+            ],
+            _source: false,
+          },
+          { ignore: [404] }
+        ),
+
+        getPendingActionsSummary(esClient, metadataService, this.log, agentIds),
+      ]).catch(catchAndWrapError);
+
+      const mostRecentAgentInfosByAgentId = searchResponse?.hits?.hits?.reduce<
+        Record<string, RawCrowdstrikeInfo>
+      >((acc, hit) => {
+        // TODO TC: check if we should use crowdstrike.cid instead
+        if (hit.fields?.['crowdstrike.host.id'][0]) {
+          acc[hit.fields?.['crowdstrike.host.id'][0]] =
+            hit.inner_hits?.most_recent.hits.hits[0]._source;
+        }
+
+        return acc;
+      }, {});
+
+      const agentStatuses = await this.getAgentStatusFromConnectorAction(agentIds);
```

Comment:
Since this may fail independently of the rest of the agent info, we should do error handling here. That way we can still show partial agent info instead of showing no status when that happens.

Reply:
I think we shouldn't show one without the other, meaning we should throw error - showing just part of the data may be confusing or even worse, may show incorrect state. What do you think?